File: 695-max-area-of-island/695-max-area-of-island.cpp
class Solution {
public:
    
    int findMaxArea(vector<vector<int>>& grid, vector<vector<bool>>& vis, int i, int j){
        
        if(i<0 || j<0 || i>=grid.size() || j>=grid[0].size()) return 0;
        if(grid[i][j] == 0 || vis[i][j] == true) return 0;
        
        vis[i][j] = true;
        
        int maxArea=0;
        // top -> left -> down -> right
        maxArea = 1 + findMaxArea(grid, vis, i-1, j) + findMaxArea(grid, vis, i, j-1) + findMaxArea(grid, vis, i+1, j) + findMaxArea(grid, vis, i, j+1);
        
        return maxArea;
    }
    
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        
        int n=grid.size();
        int m=grid[0].size();
        
        vector<vector<bool>> vis(n, vector<bool>(m, false));
        
        int maxArea=0;
        for(int i=0; i<n; i++){
            for(int j=0; j<m; j++){
                if(grid[i][j] == 1 && vis[i][j] == false){
                    maxArea = max(maxArea, findMaxArea(grid, vis, i, j));
                }
            }
        }
        
        return maxArea;
    }
};

File: 695-max-area-of-island/695-max-area-of-island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        
        rows = len(grid)
        cols = len(grid[0])
        
        visit = set()
        maxArea = 0
        
        def dfs(i, j, count):
            
            if i<0 or j<0 or i>=rows or j>=cols or grid[i][j] == 0 or (i, j) in visit:
                return
            
            count[0] += 1
            
            visit.add((i, j))
            dfs(i-1, j, count)
            dfs(i, j-1, count)
            dfs(i+1, j, count)
            dfs(i, j+1, count)
        
        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == 1 and (i, j) not in visit:
                    count = [0]
                    dfs(i, j, count)
                    maxArea = max(maxArea, count[0])
                    
        return maxArea